identity/client/src/main/java/com/azure/identity/credential/ClientCertificateCredential.java

```diff
@@ -0,0 +1,151 @@
+package com.azure.identity.credential;
+
+import com.azure.core.implementation.util.Base64Util;
+import com.microsoft.aad.adal4j.AsymmetricKeyCredential;
+import com.microsoft.aad.adal4j.AuthenticationCallback;
+import com.microsoft.aad.adal4j.AuthenticationContext;
+import com.microsoft.aad.adal4j.AuthenticationResult;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.MonoSink;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.nio.file.Files;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.Consumer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * An AAD credential that acquires a token with a client certificate for an AAD application.
+ */
+public class ClientCertificateCredential extends AadCredential<ClientCertificateCredential> {
+    private File clientCertificate;
+    private String clientCertificatePassword;
+
+    /**
+     * Creates a ClientSecretCredential with default AAD endpoint https://login.microsoftonline.com.
+     */
+    public ClientCertificateCredential() {
+        super();
+    }
+
+    /**
+     * Sets the client certificate for authenticating to AAD.
+     * @param pemCertificate the PEM file containing the certificate
+     * @return the credential itself
+     */
+    public ClientCertificateCredential pemCertificate(File pemCertificate) {
+        this.clientCertificate = pemCertificate;
+        return this;
+    }
+
+    /**
+     * Sets the client certificate for authenticating to AAD.
+     * @param pfxCertificate the password protected PFX file containing the certificate
+     * @param clientCertificatePassword the password protecting the PFX file
+     * @return the credential itself
+     */
+    public ClientCertificateCredential pfxCertificate(File pfxCertificate, String clientCertificatePassword) {
+        this.clientCertificate = pfxCertificate;
+        this.clientCertificatePassword = clientCertificatePassword;
+        return this;
+    }
+
+    @Override
+    public Mono<AuthenticationResult> authenticateAsync(String resource) {
```

Comment:
Would prefer to not have the `Async` suffix. Thoughts?

Reply:
Removed everywhere